localstack-core/localstack/aws/protocol/serializer.py

```diff
@@ -1556,25 +1589,29 @@ def _serialize_type_structure(
                 self._serialize_type_string(serialized, member_key, None, None)
                 self._serialize_data_item(serialized, member_value, member_shape)
 
+        if closing_bytes is not None:
+            serialized.extend(closing_bytes)
+
     def _serialize_type_timestamp(
         self,
         serialized: bytearray,
-        value: int | str | datetime,
+        value: int | str | datetime.datetime,
         shape: Shape | None,
         name: str | None = None,
     ) -> None:
-        timestamp = int(self._convert_timestamp_to_str(value))
+        # https://smithy.io/2.0/additional-specs/protocols/smithy-rpc-v2.html#timestamp-type-serialization
         tag = 1  # Use tag 1 for unix timestamp
         initial_byte = self._get_initial_byte(self.TAG_MAJOR_TYPE, tag)
         serialized.extend(initial_byte)  # Tagging the timestamp
-        additional_info, num_bytes = self._get_additional_info_and_num_bytes(timestamp)
 
-        if num_bytes == 0:
-            initial_byte = self._get_initial_byte(self.UNSIGNED_INT_MAJOR_TYPE, timestamp)
-            serialized.extend(initial_byte)
-        else:
-            initial_byte = self._get_initial_byte(self.UNSIGNED_INT_MAJOR_TYPE, additional_info)
-            serialized.extend(initial_byte + timestamp.to_bytes(num_bytes, "big"))
+        # we encode the timestamp as a double, like the Go SDK
+        # https://github.com/aws/aws-sdk-go-v2/blob/5d7c17325a2581afae4455c150549174ebfd9428/internal/protocoltest/smithyrpcv2cbor/serializers.go#L664-L669
+        # Currently, the Botocore serializer using unsigned integers, but it does not conform to the Smithy specs:
+        # > This protocol uses epoch-seconds, also known as Unix timestamps, with millisecond
+        # > (1/1000th of a second) resolution.
+        timestamp = float(self._convert_timestamp_to_str(value))
+        initial_byte = self._get_initial_byte(self.FLOAT_AND_SIMPLE_MAJOR_TYPE, 27)
+        serialized.extend(initial_byte + struct.pack(">d", timestamp))
```

Comment:
we might want to revisit this for compatibility reasons: while it looks like most clients should be able to deserialize this timestamp (the Go SDK deserializer is able to deserialize `float32`, `float64`, `cbor uint (within int64 bounds)` and `cbor -int (within int64 bounds)` as mentioned here: 

https://github.com/aws/smithy-go/blob/04e063a48301226f81d645e0b899f32a6455ef63/encoding/cbor/coerce.go#L145-L180

We'll need to check the AWS raw response to see what kind of resolution they return, if it depends on operation/service, and if we can have any kind of compatibility issue.